README.md
# Assignment1.04
Assignment 1.4
Please note that the Assignment is NOT COMPLETE. The current progress of the assignment is that monsters are properly generated and randomly placed in random areas in the rooms. After this they are placed in a queue where it will dequeue them and figure out how they will move. The movement for all monsters is not complete. At the moment monsters will move only if the PC is right by them thus ending the game else they won't move. The PC also does not move in this implementation. You can complile the program using make and run it the entire program using ./dun_program --nummon n; where n is the number of monster to use. 

CHANGELOG
[33mcommit 3722b7a76011c9d024f55f7f1df24de2e66a7674[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Wed Oct 4 12:39:14 2017 -0500

    only one monster can move

[33mcommit 5f99e48ea618cce046a5521fc40597257f42edb5[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Tue Oct 3 18:59:17 2017 -0500

    Queue now works and the game will end if there is a monster that is directly beside the player since the player will not move

[33mcommit 6979c0a81435baf830769e30e311792e441c4596[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Tue Oct 3 10:44:21 2017 -0500

    tryinng to move monsters

[33mcommit 370909c2828d92fc77acc4afebbc0986b18bf925[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Mon Oct 2 21:56:05 2017 -0500

    modified priority queue so monster can be properly queued

[33mcommit 73498801d9adf689f41582cbaa8e6758ffe94cb0[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Mon Oct 2 21:01:45 2017 -0500

    Monsters are placed in the dungeon in random rooms but djistras is still not working

[33mcommit 7db05b0d48f70766273bc269ca0514763f925ebc[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sun Oct 1 17:43:10 2017 -0500

    modyfing files

[33mcommit c9c77c99c4123dd22c430e613988da09f1dcbe06[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sat Sep 30 18:31:12 2017 -0500

    some code refactoring

[33mcommit f73e46eeb5470dbed13f86e227fa2d5a8fd76452[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sat Sep 30 18:21:12 2017 -0500

    Finished generating monsters now trying to move them

[33mcommit 7a935613013a47ed4644c156fb6880a50cdfb1af[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sat Sep 30 15:05:35 2017 -0500

    Finished monster generation

[33mcommit 1fa08eff83fdee5303090e3171402ab6e019cdb3[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sat Sep 30 14:18:06 2017 -0500

    adding code to handle user input and starting to generate monsters

[33mcommit 549e0fa6c7ed707e1d4ea936ebb10e697b94052c[m
Author: Juan Venegas <juanven3gas@gmail.com>
Date:   Sat Sep 30 13:06:52 2017 -0500

    Setting up project dir

[33mcommit 417b06e24dc8f67749bf6d106201ef973632d879[m
Author: Juanven3gas <juanven3gas@gmail.com>
Date:   Sat Sep 30 13:01:47 2017 -0500

    Initial commit


gen_dun.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "dun_utils.h"

int check_rooms(int rows, int columns, int start_row, int start_column);
void sort_rooms(void);
void add_rooms(void);
void add_corridors(void);
void init_dungeon(void);
void print_dungeon(void);
void palce_PC(void);
int get_pc_x_position(void);
int get_pc_y_position(void);
void init_hardness(void);

void init_hardness(void)
{
    srand(time(NULL));

    int rows, cols;

    for(rows = 0; rows < dungeon_rows; rows++)
    {
        for(cols = 0; cols < dungeon_columns; cols++)
        {
            //if we get to a point where it is a wall, corridor, or something edge then hardness is 0
            //else it is a random hardness from 1-255
            if(dungeon[rows][cols] == '.' || dungeon[rows][cols] == '#' || dungeon[rows][cols] == '|' || dungeon[rows][cols] == '-')
            {
                hardness[rows][cols] = 0;
            }
            else
            {
                int hard = (rand() % 254) + 1;
                hardness[rows][cols] = hard;
            }
        }
    }
}

void place_PC(void)
{
    srand(time(NULL));

    //Picks a random room & a random spot in the room
    //Places the PC in that random position in the dungeon
    int room = (rand() % num_rooms);
    printf("room: %d\n", room);

    int room_rows = rooms[room][0];
    int room_cols = rooms[room][1];
    int room_start_row = rooms[room][2];
    int room_start_col = rooms[room][3];

    int temp1 = (room_start_row + room_rows) -1;
    int temp2 = (room_start_col + room_cols) -1;
    
    int temp3 = room_start_row - temp1;
    int temp4 = room_start_col - temp2;

    int pc_start_row = (rand() % temp3) + room_start_row;
    int pc_start_col = (rand() % temp4) + room_start_col;

    pc_x_position = pc_start_row;
    pc_y_position = pc_start_col;
    
    dungeon[pc_start_row][pc_start_col] = '@';
}


int gen_dun(void)
{
    init_dungeon();
    add_rooms();
    add_corridors();
    init_hardness();
    //print_dungeon();

    return 0;
}


void sort_rooms(void)
{
    int i, j;

    for(i = 0; i < num_rooms - 1; i++)
    {
        int min_indx = i;

        for(j = i+1; j < num_rooms; j++)
        {
            int room1_start_y = rooms[min_indx][3];
            int room2_start_y = rooms[j][3];

            if(room2_start_y < room1_start_y)
            {
                min_indx = j;
            }

        }

        int temp_rows = rooms[min_indx][0];
        int temp_cols = rooms[min_indx][1];
        int temp_start_row = rooms[min_indx][2];
        int temp_start_cols = rooms[min_indx][3];

        rooms[min_indx][0] = rooms[i][0];
        rooms[min_indx][1] = rooms[i][1];
        rooms[min_indx][2] = rooms[i][2];
        rooms[min_indx][3] = rooms[i][3];

        rooms[i][0] = temp_rows;
        rooms[i][1] = temp_cols;
        rooms[i][2] = temp_start_row;
        rooms[i][3] = temp_start_cols;
    }
}

void add_corridors(void)
{
    int i, j, room;

    for(room = 0; room < num_rooms-1; room++)
    {
        int room1_rows = rooms[room][0];
        int room1_columns = rooms[room][1];
        int room1_start_row = rooms[room][2];
        int room1_start_column = rooms[room][3];

        int room2_rows = rooms[room+1][0];
        int room2_columns = rooms[room+1][1];
        int room2_start_row = rooms[room+1][2];
        int room2_start_column = rooms[room+1][3];

        int room1_mid_x = (room1_rows/2) + room1_start_row;
        int room1_mid_y = (room1_columns/2) + room1_start_column;

        //printf("room 1 mid x: %d, room 1 mid y: %d\n", room1_mid_x, room1_mid_y);

        int room2_mid_x = (room2_rows/2) + room2_start_row;
        int room2_mid_y = (room2_columns/2) + room2_start_column;

        //printf("room 2 mid x: %d, room 2 mid y: %d\n", room2_mid_x, room2_mid_y);

        if((room2_mid_x - room1_mid_x) < 0)
        {
            for(i =  room1_mid_x; i > room2_mid_x - 1; i --)
            {
	      if(dungeon[i][room1_start_column + room1_columns] != '.')
		{
		  dungeon[i][room1_start_column + room1_columns] = '#';
		}
                
            }
        }
        else
        {
            for(i = room1_mid_x; i < room2_mid_x + 1; i++)
            {
	      if(dungeon[i][room1_start_column + room1_columns] != '.')
		{
		   dungeon[i][room1_start_column + room1_columns] = '#';
		}
               
            }
        }

        if((room2_mid_y - room1_mid_y) < 0)
        {
            for(j = (room1_start_column + room1_columns); j > (room2_start_column) ; j--)
            {
	      if(dungeon[room2_mid_x][j] != '.')
		{
		   dungeon[room2_mid_x][j] = '#';
		}
               
            }
        }
        else
        {
            for(j = (room1_start_column + room1_columns); j < room2_start_column; j++)
            {
	      if(dungeon[room2_mid_x][j]  != '.')
		{
		  dungeon[room2_mid_x][j] = '#';
		}
                
            }
        }

    }

}

void add_rooms(void)
{
    int room_counter;
    int start_row, start_column;
    int rows, columns;
    int i, j;

    room_counter = 0;

    srand(time(NULL));

    while(room_counter < num_rooms)
    {
        int success = 1;

        //random row from 3 - 9
        rows = (rand() % 7) + 3;


        //random col from 3 - 9
        columns = (rand() % 7) + 3;

        //random row from 1 - 12
        start_row = (rand() % 11) + 1;

        //random col from 1 - 71
        start_column = (rand() % 70) + 1;

        success = check_rooms(rows, columns, start_row, start_column);

        /*
         * Add room to the dungeon & then increment the room counter
         */
        if(success)
        {
            for(i = start_row; i < (start_row + rows); i++)
            {
                for(j = start_column; j < (start_column + columns); j++)
                {
                    dungeon[i][j] = '.';
                }
            }

            rooms[room_counter][0] = rows;
            rooms[room_counter][1] = columns;
            rooms[room_counter][2] = start_row;
            rooms[room_counter][3] = start_column;

            room_counter++;
        }

    }

    sort_rooms();
    
}

int check_rooms(int rows, int columns, int start_row, int start_column)
{
    int i = 0;
    int j = 0;

    //checks to see if there is nothing above room
    for(i = (start_row - 3); i < start_row; i++)
    {
        for(j = start_column; j < (start_column + columns); j++)
        {
            if(dungeon[i][j] == '.')
            {
                return 0;
            }
        }
    }


    //checks to see if there is nothing below room

    for(i = (start_row + rows); i < (start_row + rows + 3); i++)
    {
        for(j = start_column; j < (start_column + columns); j ++)
        {
            if(dungeon[i][j] == '.')
            {
                return 0;
            }
        }
    }


    //checks to see if there is nothing to the right
    for(i = start_row; i < (start_row + rows); i++)
    {
        for(j = (start_column + columns); j < (start_column + columns + 3); j++)
        {
            if(dungeon[i][j] == '.')
            {
                return 0;
            }
        }
    }

    //checks to see if there is nothing to the left
    for(i = start_row; i < (start_row + rows); i++)
    {
        for(j = (start_column - columns); j < start_column; j++)
        {
            if(dungeon[i][j] == '.')
            {
                return 0;
            }
        }
    }


    return 1;
}

void init_dungeon(void)
{
    int i, j;

    for(i = 0; i < dungeon_rows; i++)
    {
        for(j = 0; j < dungeon_columns; j++)
        {
            dungeon[i][j] = ' ';
            dungeon[0][j] = '-';
            dungeon[dungeon_rows - 1][j] = '-';
        }

        dungeon[i][0] = '|';
        dungeon[i][dungeon_columns - 1] = '|';
    }
}

void print_dungeon(void)
{
    int i, j;

    for(i = 0; i < dungeon_rows; i++)
    {
        for(j = 0; j < dungeon_columns; j++)
        {
            printf("%c", dungeon[i][j]);
        }

        printf("\n");
    }

}

priority_queue.c
#include <stdio.h>
#include <stdlib.h>

#include "priority_queue.h"

int queue_init(queue_t *q)
{
  q->size = 0;
  //q->head = q->tail = NULL;
  q->head = (queue_node_t *)malloc(sizeof(queue_node_t));
  q->tail = (queue_node_t *)malloc(sizeof(queue_node_t));
  q->head->next = NULL;
  q->tail->next = NULL;
  return 0;
}

int queue_delete(queue_t *q)
{
  monster_t m;

  while (!queue_remove(q, &m))
    ;

  free(q->head);
  free(q->tail);
  return 0;
}

int queue_add(queue_t *q, monster_t m, int priority)
{
  //If the queue is not empty
  if (q->head->next != NULL) 
  {

    queue_node_t *toAdd = (queue_node_t *)malloc(sizeof(queue_node_t));
    
    if(toAdd == NULL)
    {
      fprintf(stderr, "Out of memory.\n");
      return 1;
    }
    /*if (!(q->tail->next = malloc(sizeof (*q->tail->next)))) {
      fprintf(stderr, "Out of memory.\n");

      return 1;
    }*/

    toAdd->monster = m;
    toAdd->priority = priority;

    queue_node_t *cursor = q->head->next;
    queue_node_t *previous = q->head;
    int addedFlag = 0;

    while(cursor->next != NULL)
    {
      if(cursor->priority > toAdd->priority)
      {
        toAdd->next = cursor;
        previous->next = toAdd;
        addedFlag = 1;
        break;
      }
      previous = cursor;
      cursor = cursor->next;
    }

    if(!addedFlag)
    {
      toAdd->next = cursor;
      previous->next = toAdd;

      /*q->tail->next = toAdd;
      toAdd->next = NULL;
      q->tail = q->tail->next;*/
    }

    /*q->tail = q->tail->next;
    q->tail->value = v;
    q->tail->next = NULL;*/
    q->size++;

    return 0;
  } 
  else 
  {
    queue_node_t *toAdd = (queue_node_t *)malloc(sizeof(queue_node_t));
    if(toAdd == NULL)
    {
      fprintf(stderr, "Out of memory.\n");
      
      return 1;
    }

    toAdd->monster = m;
    toAdd->priority = priority;
    /*if (!(q->tail = q->head = malloc(sizeof (*q->tail)))) {
      fprintf(stderr, "Out of memory.\n");

      return 1;
    }*/

    q->head->next = toAdd;
    toAdd->next= q->tail;
    /*q->tail->value = v;
    q->tail->priority = priority;
    q->tail->next = NULL;*/
    q->size++;

    return 0;
  }

  return 0; /* Won't be reached */
}

int queue_remove(queue_t *q, monster_t *m)
{
  queue_node_t *n;

  //if the queue is empty
  if (!q->size) {
    return 1;
  }

  n = q->head->next;

  q->head->next = n->next;
  q->size--;
  *m = n->monster;
  free(n);

  /*if (!q->size) {
    q->tail = NULL;
  }*/

  return 0;
}

int queue_peek(queue_t *q, monster_t *m)
{
  if (q->size) {
    *m = q->head->next->monster;

    return 0;
  }

  return 1;
}

int queue_size(queue_t *q)
{
  return q->size;
}

priority_queue.h
#ifndef PRIORITY_QUEUE_H
# define PRIORITY_QUEUE_H

typedef struct monster 
{
    int x_pos;
    int y_pos;
    int room;
    int characteristics;
    int speed;
    int move;
    int alive;
} monster_t;

typedef struct queue_node {
  //int value;
  int priority;
  monster_t monster;
  struct queue_node *next;
} queue_node_t;

typedef struct queue {
  int size;
  queue_node_t *head, *tail;
} queue_t;

int queue_init(queue_t *q);
int queue_delete(queue_t *q);
//int queue_add(queue_t *q, int v, int priority);
//int queue_remove(queue_t *q, int *v);
//int queue_peek(queue_t *q, int *v);
int queue_add(queue_t *q, monster_t m, int priority);
int queue_remove(queue_t *q, monster_t *m);
int queue_peek(queue_t *q, monster_t *m);

int queue_size(queue_t *q);

#endif

characters.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <limits.h>
#include "dun_utils.h"
#include "priority_queue.h"

void createMonsters(monster_t *arr, int size);
void placeMonsters(monster_t *arr, int size);
int moveMonster(monster_t m);
int checkForPC(monster_t m);
int charToInt(char c);
char intToChar(int i);

int game_won = 0;

int main(int argc, char* argv[])
{
    int num_monsters = 0;
    int turn = 0;

    //Check the argumens
    //and handle user inputs
    if(argc < 3)
    {
        printf("Insufficient arguements\n");
        return 1;
    }

    int arg_count = 1;
    while(arg_count < argc)
    {
        if(strcmp(argv[arg_count],  "--nummon") == 0)
        {
            arg_count++;
            num_monsters = atoi(argv[arg_count]);
            if(num_monsters <= 0)
            {
                printf("Incorrect argument passed\n");
                return 1;
            }

            //printf("Setting dungeon up for %d monsters\n", num_monsters);
            arg_count++;
        }
        else 
        {
            printf("incorrect argument detected: %s\n", argv[arg_count]);
            return 1;
        }
    }

    /**
    *End of handling user inputs and arguments
    *Start of setting up the dungeon and the distances array
    */

    calc_distances();

    /**
    * End setting up dungeon
    * Start of generating monsters with characteristics and place them in a random place in the dungeon
    */
    srand(time(NULL));

    monster_t monsters[num_monsters];
    createMonsters(monsters, num_monsters);
    placeMonsters(monsters, num_monsters);
    print_dungeon();
    
    /**
     * End of monster generation
     * Start of character movement
     */
    
     queue_t *q = (queue_t *) malloc(sizeof(queue_t));
     queue_init(q);
     int index;
     for(index = 0; index < num_monsters; index++)
     {
         int priority = event_constant / monsters[index].speed;
         queue_add(q,monsters[index], priority);
     }

     int game_finished = 0;
     
     while(!game_finished)
     {
         //Grab the first monster to move
         monster_t toMove;
         queue_remove(q, &toMove);

         if(toMove.alive)
         {
            printf("Monster %x at %d,%d will move\n", toMove.characteristics, toMove.x_pos, toMove.y_pos);
            game_finished = moveMonster(toMove);
            queue_add(q, toMove, (event_constant/toMove.speed) + turn++);
            print_dungeon();
         }
     }
     
     if(!game_won)
     {
         printf("GAME OVER PLAYER WAS EATEN BY MONSTER!!\n");
     }
     else
     {
         printf("GAME WON ALL MONSTER DESTROYED!!\n");
     }
     //print_dungeon();
     //monster_t firstToMove;
     //queue_peek(q,&firstToMove);
     //printf("The first monster to move is %x, at %d,%d\n", firstToMove.characteristics, firstToMove.x_pos, firstToMove.y_pos);
     

     queue_delete(q);
     free(q);

    return 0;
}

int moveMonster(monster_t m)
{
    
    int pcFound = checkForPC(m);

    if(pcFound)
    {
        return 1;
    }

    if(m.characteristics == 0x0)
    {
        //Look in the room to find the PC and move 1 cell closer
        int room_rows = rooms[m.room][0];
        int room_cols = rooms[m.room][1];
        int room_x = rooms[m.room][2];
        int room_y = rooms[m.room][3];

        int startX, startY;
        int pc_x = -1;
        int pc_y = -1;
        for(startX = 0; startX < (room_x + room_rows); startX++)
        {
            for(startY = 0; startY < (room_y + room_cols); startY++)
            {
                if(dungeon[startX][startY] == '@')
                {
                    pc_x = startX;
                    pc_y = startY;
                }
            }
        }

        if(pc_x == -1 || pc_y == -1)
        {
            //PC was not in the room and dont move
            return 0;
        }
        else
        {
            dungeon[m.x_pos][m.y_pos] = '.';

            if(pc_x < m.x_pos)
            {
                m.x_pos = m.x_pos - 1;
            }
            if(pc_x > m.x_pos)
            {
                m.x_pos = m.x_pos + 1;
            }
            if(pc_y < m.y_pos)
            {
                m.y_pos = m.y_pos - 1;
            }
            if(pc_y > m.y_pos)
            {
                m.y_pos = m.y_pos + 1;
            }

            dungeon[m.x_pos][m.y_pos] = '0';
        }

        return 0;
    }
    else if (m.characteristics == 0x1)
    {
        return 0;   
    }
    else if(m.characteristics == 0x2)
    {
        return 0;
    }
    else if(m.characteristics == 0x3)
    {
        //printf("Monster has a 0x3 charicteristic\n");
        int minX = 0;
        int minY = 0;
        int min = INT_MAX;
        //Get all 8 positions and see what is available
        if(distances_non_tunnel[m.x_pos+1][m.y_pos] != ' ')
        {
            if(distances_non_tunnel[m.x_pos+1][m.y_pos] == '@')
            {
                //move the monster over and win the game
                dungeon[m.x_pos+1][m.y_pos] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos+1][m.y_pos];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos+1;
                minY = m.y_pos;
            }
            
        }
        if(distances_non_tunnel[m.x_pos-1][m.y_pos] != ' ')
        {
            if(distances_non_tunnel[m.x_pos-1][m.y_pos] == '@')
            {
                dungeon[m.x_pos-1][m.y_pos] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos-1][m.y_pos];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos-1;
                minY = m.y_pos;
            }
            
        }
        if(distances_non_tunnel[m.x_pos][m.y_pos+1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos][m.y_pos+1] == '@')
            {
                dungeon[m.x_pos][m.y_pos+1] = '3';
                return 1;
            }
            
            int dis = distances_non_tunn[m.x_pos][m.y_pos+1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos;
                minY = m.y_pos+1;
            }
        }
        if(distances_non_tunnel[m.x_pos][m.y_pos-1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos][m.y_pos-1] == '@')
            {
                dungeon[m.x_pos][m.y_pos-1] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos][m.y_pos-1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos;
                minY = m.y_pos-1;
            }
        }
        if(distances_non_tunnel[m.x_pos+1][m.y_pos-1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos+1][m.y_pos-1] == '@')
            {
                dungeon[m.x_pos+1][m.y_pos-1] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos+1][m.y_pos-1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos+1;
                minY = m.y_pos-1;
            }
        }
        if(distances_non_tunnel[m.x_pos+1][m.y_pos+1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos+1][m.y_pos+1] == '@')
            {
                dungeon[m.x_pos+1][m.y_pos+1] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos+1][m.y_pos+1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos+1;
                minY = m.y_pos+1;
            }
        }
        if(distances_non_tunnel[m.x_pos-1][m.y_pos+1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos-1][m.y_pos+1] == '@')
            {
                dungeon[m.x_pos-1][m.y_pos+1] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos-1][m.y_pos+1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos-1;
                minY = m.y_pos+1;
            }
        }
        if(distances_non_tunnel[m.x_pos-1][m.y_pos-1] != ' ')
        {
            if(distances_non_tunnel[m.x_pos-1][m.y_pos-1] == '@')
            {
                dungeon[m.x_pos-1][m.y_pos-1] = '3';
                return 1;
            }
            int dis = distances_non_tunn[m.x_pos-1][m.y_pos-1];
            if(dis < min)
            {
                min = dis;
                minX = m.x_pos-1;
                minY = m.y_pos-1;
            }
            
        }

        dungeon[minX][minY] = '3';
        dungeon[m.x_pos][m.y_pos] = '.';
        m.x_pos = minX;
        m.y_pos = minY;
        return 0;
    }
    else if(m.characteristics == 0x4)
    {
        return 0;
    }
    else if(m.characteristics == 0x5)
    {
        return 0;
    }
    else if(m.characteristics == 0x6)
    {
        return 0;
    }
    else if(m.characteristics == 0x7)
    {
        return 0;
    }
    else if(m.characteristics == 0x8)
    {
        return 0;
    }
    else if(m.characteristics == 0x9)
    {
        return 0;
    }
    else if (m.characteristics == 0xa)
    {
        return 0;
    }
    else if (m.characteristics == 0xb)
    {
        return 0;   
    }
    else if (m.characteristics == 0xc)
    {
        return 0;
    }
    else if (m.characteristics == 0xd)
    {
        return 0;
    }
    else if (m.characteristics == 0xe)
    {
        return 0;
    }
    else if (m.characteristics == 0xf)
    {
        return 0;
    }
    
    return 0;
}

int checkForPC(monster_t m)
{
    if(dungeon[m.x_pos +1][m.y_pos] == '@')
    {
        dungeon[m.x_pos+1][m.y_pos] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos -1][m.y_pos] == '@')
    {
        dungeon[m.x_pos-1][m.y_pos] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos][m.y_pos+1] == '@')
    {
        dungeon[m.x_pos][m.y_pos+1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos][m.y_pos-1] == '@')
    {
        dungeon[m.x_pos][m.y_pos-1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos +1][m.y_pos+1] == '@')
    {
        dungeon[m.x_pos+1][m.y_pos+1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos +1][m.y_pos-1] == '@')
    {
        dungeon[m.x_pos+1][m.y_pos-1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos -1][m.y_pos+1] == '@')
    {
        dungeon[m.x_pos-1][m.y_pos+1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }
    else if(dungeon[m.x_pos -1][m.y_pos-1] == '@')
    {
        dungeon[m.x_pos-1][m.y_pos-1] = intToChar(m.characteristics);
        dungeon[m.x_pos][m.y_pos] = '.';
        return 1;
    }

    return 0;
}

void placeMonsters(monster_t *arr, int size)
{
    int index;
    srand(time(NULL));

    for(index = 0; index < size; index++)
    {
        //Pick a random room
        int room = rand() % num_rooms;

        int room_rows = rooms[room][0];
        int room_cols = rooms[room][1];
        int room_start_row = rooms[room][2];
        int room_start_col = rooms[room][3];

        int temp1 = (room_start_row + room_rows) -1;
        int temp2 = (room_start_col + room_cols) -1;
        
        int temp3 = room_start_row - temp1;
        int temp4 = room_start_col - temp2;
    
        //Pick a random position in that room so long as there is nothing there
        int success = 0;
        while(!success)
        {
            int monster_start_row = (rand() % temp3) + room_start_row;
            int monster_start_col = (rand() % temp4) + room_start_col;

            if(dungeon[monster_start_row][monster_start_col] == '.')
            {
                //place the character of the monster
                int characteristics = arr[index].characteristics;
                arr[index].x_pos = monster_start_row;
                arr[index].y_pos = monster_start_col;
                arr[index].room = room;
                success = 1;

                switch (characteristics)
                {
                    case 0x0 :
                    dungeon[monster_start_row][monster_start_col] = '0';
                    break;
                    case 0x1 :
                    dungeon[monster_start_row][monster_start_col] = '1';
                    break;
                    case 0x2 :
                    dungeon[monster_start_row][monster_start_col] = '2';
                    break;
                    case 0x3 :
                    dungeon[monster_start_row][monster_start_col] = '3';
                    break;
                    case 0x4 :
                    dungeon[monster_start_row][monster_start_col] = '4';
                    break;
                    case 0x5 :
                    dungeon[monster_start_row][monster_start_col] = '5';
                    break;
                    case 0x6 :
                    dungeon[monster_start_row][monster_start_col] = '6';
                    break;
                    case 0x7 :
                    dungeon[monster_start_row][monster_start_col] = '7';
                    break;
                    case 0x8 :
                    dungeon[monster_start_row][monster_start_col] = '8';
                    break;
                    case 0x9 :
                    dungeon[monster_start_row][monster_start_col] = '9';
                    break;
                    case 0xa :
                    dungeon[monster_start_row][monster_start_col] = 'a';
                    break;
                    case 0xb :
                    dungeon[monster_start_row][monster_start_col] = 'b';
                    break;
                    case 0xc :
                    dungeon[monster_start_row][monster_start_col] = 'c';
                    break;
                    case 0xd :
                    dungeon[monster_start_row][monster_start_col] = 'd';
                    break;
                    case 0xe :
                    dungeon[monster_start_row][monster_start_col] = 'e';
                    break;
                    case 0xf :
                    dungeon[monster_start_row][monster_start_col] = 'f';
                    break;
                    default :
                    dungeon[monster_start_row][monster_start_col] = 'g';
                }
            }
        }
    }
}

void createMonsters(monster_t *arr, int size)
{
    int index;
    for(index = 0; index < size; index++)
    {
        int characteristics = rand() & 0xf;
        int speed = (rand() % 15) + 5;
        arr[index].characteristics = characteristics;
        arr[index].speed = speed;
        arr[index].move = event_constant / speed;
        arr[index].alive = 1;
        //printf("Characteristics for monster %d is %x and speed is %d and is moving in turn %d\n", index, arr[index].characteristics, arr[index].speed, arr[index].move);
    }
}

int charToInt(char c)
{
    switch(c)
    {
        case '0':
        return 0;
        case '1':
        return 1;
        case '2':
        return 2;
        case '3':
        return 3;
        case '4':
        return 4;
        case '5':
        return 5;
        case '6':
        return 6;
        case '7':
        return 7;
        case '8':
        return 8;
        case '9':
        return 9;
    }

    return 0;
}

char intToChar(int i)
{
    switch(i)
    {
        case 0:
        return '0';
        case 1:
        return '1';
        case 2:
        return '2';
        case 3:
        return '3';
        case 4:
        return '4';
        case 5:
        return '5';
        case 6:
        return '6';
        case 7:
        return '7';
        case 8:
        return '8';
        case 9:
        return '9';
        case 10:
        return 'a';
        case 11:
        return 'b';
        case 12:
        return 'c';
        case 13:
        return 'd';
        case 14:
        return 'e';
        case 15:
        return 'f';
    }

    return '0';
}